Zac/create prayers

zbarnard1 opened the pull request:
Add route to POST Prayers

This route creates a new Prayer for the User and includes the ID of the next Resident in the list to minimize the number of requests we need to make.

There is a new validation on Prayers which ensures that we aren't trying to save one without a Resident and raising a database-level error, resulting in a 500 status. This is a client error,  so we need to return a 4xx status.

zbarnard1 committed:
Add route to POST Prayers

This route creates a new Prayer for the User and
includes the ID of the next Resident in the list
to minimize the number of requests we need to make.

There is a new validation on Prayers which ensures
that we aren't trying to save one without a
Resident and raising a database-level error,
resulting in a 500 status. This is a client error,
so we need to return a 4xx status.

zbarnard1 commented on spec/models/prayer_spec.rb:
```diff
@@ -11,7 +11,7 @@
     it "must have a resident" do
       expect do
         build(:prayer, resident: nil).save!
-      end.to raise_error(ActiveRecord::NotNullViolation)
```
@ittullos I noticed we've been expecting the wrong error here, this is going to give us a 500 status code on a request, we should be using a validation to prevent this error from happening.

ittullos replied:
Gotcha!

zbarnard1 commented on app/models/prayer.rb:
```diff
@@ -6,4 +6,5 @@ class Prayer < ActiveRecord::Base
   belongs_to :route, optional: true
 
   validates :recorded_at, presence: true
+  validates :resident, presence: true
```
This gives us a better error when the resident_id is missing from a POST request to /prayers.

ittullos replied:
Ok yeah I probably should have added a validation here

zbarnard1 commented on spec/requests/prayers_spec.rb:
```diff
@@ -0,0 +1,66 @@
+      )
+    end
+
+    it 'returns an error response for a malformed request' do
```
I might change this test a bit because it doesn't test what it says it does. If the request had invalid JSON then I would consider it malformed. This is just an incorrect parameter, but valid JSON.

zbarnard1 commented on app/app.rb:
```diff
@@ -78,3 +78,26 @@
+  prayer = Prayer.new(resident_id:, user_id: user&.id, recorded_at: Time.current)
+
+  content_type :json
+  return [400, prayer.errors] unless prayer.save!
```
I know a better way to do this, but I might wait for the next endpoint to do any refactoring. 🤔

ittullos commented on app/app.rb:
```diff
@@ -78,3 +78,29 @@
+post '/prayers' do
+  email = request.fetch_header('P4L-email')
+  user = User.find_by_email(email)
+
```
Should this be changed to match what the header looks like coming from React Native?

When I send the header ("P4L-email") from the front end it changes the key to "HTTP_P4L_EMAIL".
Or is this maybe a problem with how I'm sending the headers?

zbarnard1 replied:
You're absolutely right. I'm getting ready to re-work the whole auth thing so I just didn't want to mess with it yet.
Why don't I fix this on the next PR where I add the auth middleware?

ittullos replied:
Sounds good to me!! Just thought I would mention it

zbarnard1 committed:
Add Authentication middleware

This middleware sits in front of every endpoint and will retreive the
User based on the email header, which allows us to remove all the
duplicated code that used to do this.
The entire User object is stored in the request.env for access within
the action.
For whatever reason, the app.rb file is in scope when Factorybot runs,
so I've had to give the convenience method for the user a longer name
to avoid strange errors during test setup.

This also changes the header to something more realistic. When the
header originated from a real client, it's going to be all caps and
prefixed with `HTTP_`.

ittullos committed:
Merge pull request #15 from ittullos/zac/add-auth-middleware

Add Authentication middleware
